pageserver/src/tenant/blob_io.rs

```diff
@@ -66,12 +69,29 @@ impl<'a> BlockCursor<'a> {
                 len_buf.copy_from_slice(&buf[off..off + 4]);
                 off += 4;
             }
-            len_buf[0] &= 0x7f;
+            len_buf[0] &= 0x0f;
             u32::from_be_bytes(len_buf) as usize
         };
+        let compression_bits = first_len_byte & 0xf0;
 
-        dstbuf.clear();
-        dstbuf.reserve(len);
+        let mut tmp_buf = Vec::new();
+        let buf_to_write;
+        let compression = if compression_bits <= BYTE_UNCOMPRESSED {
```

Comment:
> reserve the most significant byte in the length for compression

In general only the most significant 4 bits are reserved, so half a byte, of which the top bit is already reserved, so only 3 additional bits are being reserved by this PR.

Of the 7 possible states the three bits encode, only 1 is taken by this PR for compression. Probably another state would be useful for "future bytes", so there is 5 unassigned but reserved possible patterns of bits. 

Yes, previously these patterns were used to encode blobs of large sizes, but I've written in the PR description that this PR is filed under the assumption that these large sizes are not present. At least in grafana, the warning that PR #7852 has added has never shown up. If blobs of such large sizes *are* present, we'll now get errors when writing or reading them. It was John's idea to use unrealistic lengths for encoding compression instead of adding additional bytes, see https://github.com/neondatabase/neon/pull/7091#discussion_r1521750331

Reply:
Okay, makes sense. I'm okay with this, but let's leave a comment in this code about this decision to take some bits away from the size.

Reply:
added a comment in the constants section.